Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift

```diff
@@ -0,0 +1,263 @@
+    struct StateMachine: Sendable {
+        /// The ``NIOAsyncSequenceProducer.Source`` used for yielding the messages to the ``NIOAsyncSequenceProducer``.
+        var sequenceSource: Producer.Source? // TODO: make sendable?
+
+        /// The events that can be triggered by the ``KafkaBackPressurePollingSystem``.
+        enum Event {
+            /// Produce more elements.
+            case produceMore
+            /// Suspend the poll loop until the continuation is resumed.
+            case suspendLoop(CheckedContinuation<Void, Never>)
+            /// Our `Task` was cancelled why we were waiting for the loop to be unsuspended.
+            case loopSuspensionCancelled
+            /// Request to stop producing elements. When this reaches the poll loop it will escalate
+            /// to a full suspension.
+            case stopProducing
+            /// Close the entire ``KafkaBackPressurePollingSystem`` with its poll loop.
+            case shutdown
+        }
+
+        /// The possible states of the state machine.
+        enum State {
+            /// Initial state.
+            case initial
+            /// The system up and producing acknowledgement messages.
+            case producing
+            /// The pool loop is currently suspended and we are waiting for an invocation
+            /// of `produceMore()` to continue producing messages.
+            case stopProducing(CheckedContinuation<Void, Never>?)
+            /// The system is shut down.
+            case finished
+        }
+
+        /// The current state of the state machine.
+        var state = State.initial
+
+        /// The possible actions for the poll loop.
+        enum PollLoopAction {
+            /// Ask `librdkakfa` to receive new message acknowledgements at a given poll interval.
+            case pollAndSleep
+            /// Suspend the poll loop.
+            case suspendPollLoop
+            /// Shutdown the poll loop.
+            case shutdownPollLoop
+        }
+
+        /// Determines the next action to be taken in the poll loop based on the current state.
+        ///
+        /// - Returns: The next action for the poll loop.
+        func nextPollLoopAction() -> PollLoopAction {
+            switch self.state {
+            case .initial, .producing:
+                return .pollAndSleep
+            case .stopProducing:
+                // We were asked to stop producing,
+                // but the poll loop is still running.
+                // Trigger the poll loop to suspend.
+                return .suspendPollLoop
+            case .finished:
+                return .shutdownPollLoop
+            }
+        }
+
+        /// Our downstream consumer allowed us to produce more elements.
+        mutating func produceMore() {
+            switch self.state {
+            case .finished, .producing:
+                return
+            case .stopProducing(let continuation):
+                continuation?.resume()
+                fallthrough
```

Comment:
Please avoid using `fallthrough` it makes the code harder to reason about. Rather prefer explicit state transition in every case.